src/calibre/utils/speedup.c

```diff
@@ -128,23 +128,25 @@ speedup_detach(PyObject *self, PyObject *args) {
 
 static PyObject*
 speedup_fdopen(PyObject *self, PyObject *args) {
-    PyObject *ans = NULL, *name = NULL;
-    PyFileObject *t = NULL;
+    PyObject *ans = NULL;
+    char *name;
+#if PY_MAJOR_VERSION == 2
     FILE *fp = NULL;
+#endif
     int fd = -1, bufsize = -1;
     char *mode = NULL;
 
-    if (!PyArg_ParseTuple(args, "iOs|i", &fd, &name, &mode, &bufsize)) return NULL;
+    if (!PyArg_ParseTuple(args, "iss|i", &fd, &name, &mode, &bufsize)) return NULL;
+#if PY_MAJOR_VERSION >= 3
+    ans = PyFile_FromFd(fd, name, mode, bufsize, NULL, NULL, NULL, 1);
+#else
+    ans = PyFile_FromFile(fp, name, mode, fclose);
     fp = fdopen(fd, mode);
```

Comment:
should come before pyfile_fromfile not after it

Reply:
I was debugging this right before I went to bed, I have no idea how I missed it :)